### ISSUE TITLE
Gradients are not averaged when using GradientDescentOptimizer

### ISSUE THREAD

**Issue by theflofly** [tensorflow/tensorflow#8746]
When using the GradientDescentOptimizer the weights are updated using `var.device(d) -= grad * lr();`. The gradients are not averaged as we could expect. That why my model was working fine with a mini-batch gradient descent but not with the batch gradient descent.

The trick is to use `GradientDescentOptimizer(1/m * 0.01)` where `m` is the size of the batch.

The corresponding StackOverflow issue is here: http://stackoverflow.com/questions/43027109/tensorflow-weights-increasing-when-using-the-full-dataset-for-the-gradient-desce

**Comment by MicaelCarvalho** [comment]
The problem isn't the Optimizer, it's your loss. It should return the mean loss, not the sum. If you're doing an L2 regression, for instance, it should look like this:

```
l_value = tf.pow(tf.abs(ground_truth - predict), 2) # distance for each individual position of the output matrix of shape = (n_examples, example_data_size)
regression_loss = tf.reduce_sum(l_value, axis=1, name='regression_loss') # distance per example, shape = (n_examples, 1)
total_regression_loss = tf.reduce_mean(regression_loss, name='total_regression_loss') # mean distance of all examples, shape = (1)
```

**Comment by theflofly** [comment]
My bad (facepalm).
